Documentation/teaching/so2/assign2-driver-uart.rst

```diff
@@ -0,0 +1,163 @@
+
+Statement
+=========
+
+Write a kernel module that implements a driver for the serial port (`UART16550`).
+The device driver must support the two standard serial ports in a PC, `COM1` and `COM2` (`0x3f8` and `0x2f8`,
+in fact the entire range of `8` addresses `0x3f8-0x3ff` and `0x2f8-0x2ff` specific to the two ports).
+In addition to the standard routines (`open`, `read`, `write`, `close`),
+the driver must also have support for changing communication parameters using an `ioctl` operation (`UART16550_IOCTL_SET_LINE`).
+
+The driver must use interrupts for both reception and transmission to reduce latency and CPU usage time.
+`Read` and `write` calls must also be blocking. :command:`Assignments that do not meet these requirements will not be considered.`
+It is recommended that you use a buffer for read and another buffer for write buffer for each serial port in the driver.
+
+A blocking read call means that the read routine called from the user-space will be blocked until :command:`at least` one byte is read
+(the read buffer in the kernel is empty and no data can be read).
+A blocking write call means that the write routine called from the user-space will be blocked until :command:`at least` one byte is written
+(the write buffer in the kernel is full and no data can be written).
+
+Buffers Scheme
+--------------
+
+.. image:: ../img/buffers-scheme.png
+
+Data transfer between the various buffers is a `Producer-Consumer <https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem>`__ problem. Example:
+
+-   if it is written from a process to the device, the process is the producer and the device is the consumer; the process will block until there is at least one free space in the consumer's buffer
+
+-   if it is read from a process from the device, the process is the consumer and the device the producer; the process will block until there is at least one element in the producer's buffer.
+
+Implementation Details
+======================
+
+-  the driver will be implemented as a kernel module named :command:`uart16550.ko`
+-  the driver will be accessed as a character device driver, with different functions depending on the parameters transmitted to the load module:
+
+   -  the `major` parameter will specify the major with which the device must be registered
+   -  the `option` parameter will specify how it works:
+
+      -  OPTION_BOTH: will also register COM1 and COM2, with the major given by the `major` parameter and the minors 0 (for COM1) and 1 (for COM2);
+      -  OPTION_COM1: will only register COM1, with the major `major` and minor 0;
+      -  OPTION_COM2: will only register COM2, with the major `major` and minor 1;
+   -  to learn how to pass parameters in Linux, see `tldp <https://tldp.org/LDP/lkmpg/2.6/html/x323.html>`__
+   -  the default values are `major=42` and `option=OPTION_BOTH`.
+-  the number of interrupts associated with COM1 is 4 (`IRQ_COM1`) and the number of interrupts associated with COM2 is 3 (`IRQ_COM2`)
```

Comment:
Here it's actually the "interrupt number".
So it becomes: - the interrupt number associated with COM1 is 4 ...